core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/EntityVariableUpdaterLookup.java

```diff
@@ -4,63 +4,124 @@
                         if (sharedValue.getValue() == null) {
                             sharedValue.setValue(valueSupplier.get());
                         }
                     });
         };
-        return new EntityVariableUpdaterLookup<>(lookupSupplier);
+        return new EntityVariableUpdaterLookup<>(lookupSupplier,
+                (entity, variableMetamodel) -> variableMetamodel);
     }
 
     public static <Solution_> EntityVariableUpdaterLookup<Solution_> entityDependentLookup() {
         Supplier<Lookup<Solution_>> lookupSupplier = () -> {
             var valueMap = new IdentityHashMap<Object, List<VariableUpdaterInfo<Solution_>>>();
-            return new Lookup<>(valueMap::get,
-                    (entity, valueSupplier) -> valueMap.computeIfAbsent(entity, ignored -> valueSupplier.get()));
+            return new Lookup<>((entity, ignored) -> valueMap.get(entity),
+                    (entity, ignored, valueSupplier) -> valueMap.computeIfAbsent(entity, (ignored2) -> valueSupplier.get()));
         };
-        return new EntityVariableUpdaterLookup<>(lookupSupplier);
+        return new EntityVariableUpdaterLookup<>(lookupSupplier,
+                (entity, variableMetamodel) -> variableMetamodel);
+    }
+
+    public static <Solution_> EntityVariableUpdaterLookup<Solution_>
+            groupedEntityDependentLookup(
+                    Function<VariableMetaModel<Solution_, ?, ?>, @Nullable Function<Object, @Nullable Object>> variableToGroupKeyMapper) {
+        Supplier<Lookup<Solution_>> lookupSupplier = () -> {
+            var valueMap = new IdentityHashMap<Object, List<VariableUpdaterInfo<Solution_>>>();
+            var groupValueMap = new LinkedHashMap<Object, List<VariableUpdaterInfo<Solution_>>>();
+
+            BiFunction<Object, VariableMetaModel<Solution_, ?, ?>, List<VariableUpdaterInfo<Solution_>>> valueReader =
+                    (entity, variableMetaModel) -> {
+                        var groupMapper = variableToGroupKeyMapper.apply(variableMetaModel);
+                        if (groupMapper != null) {
+                            var groupKey = groupMapper.apply(entity);
+                            if (groupKey != null) {
+                                return groupValueMap.get(groupKey);
+                            }
+                        }
+                        return valueMap.get(entity);
+                    };
+
+            TriConsumer<Object, VariableMetaModel<Solution_, ?, ?>, Supplier<List<VariableUpdaterInfo<Solution_>>>> valueSetter =
+                    (entity, variableMetaModel, valueSupplier) -> {
+                        var groupMapper = variableToGroupKeyMapper.apply(variableMetaModel);
+                        if (groupMapper != null) {
+                            var groupKey = groupMapper.apply(entity);
+                            if (groupKey != null) {
+                                groupValueMap.computeIfAbsent(groupKey, ignored -> valueSupplier.get());
```

Comment:
If this is called often enough on the hot path, computeIfAbsent will create a capturing lambda every time it is called. The good old get/put-if-null works better in these cases.